### PR TITLE
XorT fromXor, fromTryCatch, and fromTry

### PR DESCRIPTION
**Pull request by adelbertc**

### CODE REVIEW

**Comment by non** [comment]
:+1:

File: tests/src/test/scala/cats/tests/XorTTests.scala [hunk 1]
```diff
@@ -23,4 +27,34 @@ class XorTTests extends CatsSuite {
...
+    }
+  }
+
+  test("fromXor")(check {
+    forAll { (xor: Xor[String, Int]) =>
+      Some(xor.isLeft) == XorT.fromXor[Option](xor).isLeft
+    }
+  })
+
+  implicit val arbitraryTryInt: Arbitrary[Try[Int]] = Arbitrary {
```
Comment (rossabaker) [inline tests/src/test/scala/cats/tests/XorTTests.scala]: We're starting to repeat `Arbitrary[Try[_]]`.  While we wait on rickynils/scalacheck#170, should we centralize that somewhere for our own use?

**Comment by fthomas** [comment]
@adelbertc I'm curious about those *Aux classes. What is their purpose?

**Comment by adelbertc** [comment]
@fthomas - The *Aux classes are to help inference and let users only specify the unknown. If we had made it:

```scala
def fromTry[F[_], A](t: Try[A]): XorT[F, Throwable, A] = ...
```

Then users would have to do

```scala
val t: Try[Int] = ...
fromTry[Option, Int](t)
```

Since we can't specify only part of the type parameters. The `Aux` classes allow us to do:

```scala
fromTry[Option] // now a FromTryAux[Option]
fromTry[Option](t) // Calls FromTryAux[Option]#apply[A], inferring A to be Int
```

@rossabaker I can make a trait in the tests called `OrphanInstances` or something and mix it in with the `CatsSuite` ?

**Comment by fthomas** [comment]
Ah, yes currying type parameters. Interesting that I haven't recognized it, because I've used this trick myself recently. One downside of this is that the Scaladoc for those functions won't be very helpful (especially if they return private types). Do you think it makes sense to document the purpose of the *Aux classes on the `fromTry`, `fromTryCatch`, and `fromXor` or how those functions should be used?

**Comment by non** [comment]
Just because a bunch has happened, I will restate my :+1:

File: core/src/main/scala/cats/data/XorT.scala [hunk 2]
```diff
@@ -107,12 +111,66 @@ case class XorT[F[_], A, B](value: F[A Xor B]) {
...
   final def left[F[_], A, B](fa: F[A])(implicit F: Functor[F]): XorT[F, A, B] = XorT(F.map(fa)(Xor.left))
 
   final def right[F[_], A, B](fb: F[B])(implicit F: Functor[F]): XorT[F, A, B] = XorT(F.map(fb)(Xor.right))
 
   final def pure[F[_], A, B](b: B)(implicit F: Applicative[F]): XorT[F, A, B] = right(F.pure(b))
+
+  /** Transforms an `Xor` into an `XorT`, lifted into the specified `Applicative`.
+   *
+   * Note: The return type is a FromXorAux[F], which has an apply method on it, allowing
+   * you to call fromTry like this:
```
Comment (ceedubs) [inline core/src/main/scala/cats/data/XorT.scala]: copy/paste error :)

**Comment by fthomas** [comment]
@adelbertc Thanks for adding documentation!

File: core/src/main/scala/cats/data/XorT.scala [hunk 3]
```diff
@@ -1,6 +1,10 @@
 package cats
 package data
 
+import scala.reflect.ClassTag
+
+import scala.util.Try
+
```
Comment (fthomas) [inline core/src/main/scala/cats/data/XorT.scala]: Those imports are not needed any more, right?

File: tests/src/test/scala/cats/tests/XorTTests.scala [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
-import cats.data.XorT
+import cats.data.{Xor, XorT}
 import cats.laws.discipline.{MonadErrorTests, MonoidKTests, SerializableTests}
 import cats.laws.discipline.arbitrary._
 
+import org.scalacheck.{Arbitrary, Gen}
+import org.scalacheck.Arbitrary.arbitrary
 import org.scalacheck.Prop.forAll
 
+import scala.util.{Failure, Success, Try}
```
Comment (ceedubs) [inline tests/src/test/scala/cats/tests/XorTTests.scala]: Not a big deal, but I think there are some extra imports lingering here too.

**Comment by ceedubs** [comment]
There are some unnecessary imports which it might be nice to remove, but aren't a big deal. :+1:

**Pull request merged**